feat: allow round trip middleware per operation

pseudo-su opened the pull request:
Reference implementation to cater to the use-cases I described in #150


## Notes

Fixes https://github.com/deepmap/oapi-codegen/issues/150

StevenACoffman commented:
This is exactly the use case I have, so I would love to see a complete example in your pseudo-su /golang-service-template repo

pseudo-su commented:
@StevenACoffman I haven't forgotten about this, you make a good point about examples... I might actually add some more tests cases to this PR that would also help somewhat in capturing examples.

I have a few open PRs for `oapi-codegen` that may or may not get merged. I was trying to avoid it because I'm lazy but I might merge them all into my forked repo and add a branch to `pseudo-su /golang-service-template` with examples.

My preference was to get things merged (where possible) before spending too much time messing about with a temporary fork.

StevenACoffman commented on pkg/codegen/templates/client.tmpl:
```diff
@@ -42,9 +96,71 @@ func NewClient(server string, opts ...ClientOption) (*Client, error) {
     return &client, nil
 }
 
+func setupoperationDoers(c *Client, rtMiddlewares RoundTripMiddlewares) *operationDoFunctions {
```
The case on this is weird. 
```suggestion
func setupOperationDoers(c *Client, rtMiddlewares RoundTripMiddlewares) *operationDoFunctions {
```

StevenACoffman commented on pkg/codegen/templates/client.tmpl:
```diff
@@ -42,9 +96,71 @@ func NewClient(server string, opts ...ClientOption) (*Client, error) {
         client.Client = http.DefaultClient
     }
+
+    client.operationDoers = setupoperationDoers(&client, client.RoundTripMiddlewares)
```
```suggestion
    client.operationDoers = setupOperationDoers(&client, client.RoundTripMiddlewares)
```

StevenACoffman commented:
@deepmap-marcinr This one is ready too!

StevenACoffman commented:
@pseudo-su Hmmm, I think your rebase you need to regenerate because now there's some test failures.

pseudo-su commented:
I need to fix some things and add those example/tests I mentioned

pseudo-su commented:
Ok @StevenACoffman three main things fixed and this should be good now.

* joinMiddleware had an issue when passing an empty list of middleware to join.
* I've added tests in `internal/test/client/client_test.go` that test an "interceptor" using `WithSharedRoundTripMiddleware` and `WithRoundTripMiddleware`
* I fixed the spec in `internal/test/client/client.yaml` because the responses were missing the `content` field inbetween the status code and the content-type

EG this
```yaml
responses:
  200:
    application/json:
```
instead of 
```yaml
responses:
  200:
    content:
      application/json:
```

StevenACoffman commented:
Thanks, yes, this works for us in our client. Nice improvement! @deepmap-marcinr Ready for you!

batazor commented:
What is the status of this PR? Is there any other way to set custom behavior?

@StevenACoffman @pseudo-su

DAcodedBEAT commented:
@oapi-codegen/oapi-codegen any reason why this wasn't approved/merged (aside from the current merge conflicts)?

DAcodedBEAT commented:
@oapi-codegen any reason why this wasn't approved/merged (aside from the current merge conflicts)?
